### PR TITLE
feat(property-values): speed up property searches by up to 150x

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Problem

#13739 

We were doing a full search over all events for a team when doing property value searches e.g. for autocomplete.

For <our_biggest_customer>, that meant things like autocomplete queries taking 50s. A separate problem is that on every keystroke we run another one of those queries :D But that isn't fixed here.

## Changes

Allow passing multiple `event_name` query params to the `event/values` endpoint which we'll use to filter on the query to look for distinct property values. 

For one query <our_biggest_customer> was running, this improves things 150x 🤯 
Screenshot 2023-01-16 at 12 13 28

## How did you test this code?

Added tests, tested manually.

### CODE REVIEW

**Comment by macobo** [comment]
Q on the benchmark: Did you control for page cache?

This looks good to me, but I'd like to capture timing data before we merge this. PR incoming...

~~Also note (given this was the sprint goal) I was thinking of benchmarking moving this logic to a separate mv completely. However this should not affect merging this PR.~~ Nevermind, I was thinking of the property keys rather than values.

File: posthog/queries/property_values.py [hunk 1]
```diff
@@ -1,34 +1,43 @@
...
+):
     property_field, _ = get_property_string_expr("events", key, "%(key)s", "properties")
     parsed_date_from = "AND timestamp >= '{}'".format(relative_date_parse("-7d").strftime("%Y-%m-%d 00:00:00"))
     parsed_date_to = "AND timestamp <= '{}'".format(timezone.now().strftime("%Y-%m-%d 23:59:59"))
+    event_filter = ""
+    value_filter = ""
+    value_params = {}
+
+    if event_names:
+        event_conditions = " OR ".join([f"event = '{event_name}'" for event_name in event_names])
```
Comment (macobo) [inline posthog/queries/property_values.py:24]: Nit: No need for creating a list - you can drop []

File: posthog/queries/property_values.py [hunk 2]
```diff
@@ -1,34 +1,43 @@
...
+    key: str, team: Team, event_names: Optional[List[str]] = None, value: Optional[str] = None
+):
     property_field, _ = get_property_string_expr("events", key, "%(key)s", "properties")
     parsed_date_from = "AND timestamp >= '{}'".format(relative_date_parse("-7d").strftime("%Y-%m-%d 00:00:00"))
     parsed_date_to = "AND timestamp <= '{}'".format(timezone.now().strftime("%Y-%m-%d 23:59:59"))
+    event_filter = ""
+    value_filter = ""
+    value_params = {}
+
+    if event_names:
```
Comment (macobo) [inline posthog/queries/property_values.py:23]: Nit: Let's be explicit about null and empty array handling.

**Review by mariusandra — APPROVED** [review]
Looks good to me, though I'd do what Karl says 😅

File: posthog/models/event/sql.py [hunk 3]
```diff
@@ -271,9 +259,11 @@
     events
 WHERE
     team_id = %(team_id)s AND
-    {property_field} ILIKE %(value)s
+    JSONHas(properties, %(key)s)
```
Comment (macobo) [inline posthog/models/event/sql.py:262]: Q: Could we rely on materialized columns here when possible?

**Comment by yakkomajuri** [comment]
Benchmarked without page cache, here are the results:

Screenshot 2023-01-17 at 11 24 58

The key thing is how much we reduce the number of rows read by filtering on `event` which is part of our sorting key. This is particularly important when the number of returned rows < limit.

File: posthog/models/event/sql.py (hunk 3 above)
Comment (yakkomajuri) [reply posthog/models/event/sql.py]: We were already doing so, but this check didn't need to be there for mat columns. Seems it was a bug that was also there before, fixed.

**Comment by yakkomajuri** [comment]
All sorted!

File: posthog/queries/property_values.py [hunk 4]
```diff
@@ -1,34 +1,56 @@
...
 
     if value:
-        return insight_sync_execute(
-            SELECT_PROP_VALUES_SQL_WITH_FILTER.format(
-                parsed_date_from=parsed_date_from, parsed_date_to=parsed_date_to, property_field=property_field
-            ),
-            {"team_id": team.pk, "key": key, "value": "%{}%".format(value)},
-            query_type="get_property_values_with_value",
-        )
+        value_filter = "AND {} ILIKE %(value)s".format(property_field)
```
Comment (macobo) [inline posthog/queries/property_values.py:40]: Nit: prefer f-strings

**Review by macobo — APPROVED** [review]

**Pull request merged**